### DISCUSSION TITLE
USB hot-plugging

### DISCUSSION THREAD

**Discussion by martinwhitaker** [memtest86plus/memtest86plus#182] [Enhancement]
Currently the USB drivers do not support hot-plugging, so when using the built-in USB drivers, any USB keyboards must be plugged in before starting Memtest86+ and must remain plugged in for the duration of the test if they are to remain usable. Plugging or unplugging USB devices whilst Memtest86+ is running should not cause any errors - it simply won't notice them.

This was a deliberate design decision, both to keep the code simple and to minimise the amount of memory permanently allocated to the USB controllers and drivers (which is then excluded from the memory tests). If it were decided to add support for hot-plugging in the future, I would suggest making it a boot option.

**Comment by debrouxl** [comment]
If such support were to be added, I guess that it should be made a boot option, indeed. As I understand it, on my computers, the amount of memory permanently set aside from the tested range should be a 4K page for OHCI, UHCI and EHCI controllers; XHCI controllers can have larger mappings, though.

I didn't think that USB device hotplugging was an important feature for a program like memtest86+; #179 has slightly changed my mind on that front. BMCs do provide important functionality for monitoring or remote access, at the expense of security: this kind of low-level, complex software needs frequent patches for critical vulnerabilities. I disable and do not connect them on "my" several servers, but that does not scale...
